### PR TITLE
gh-87901: os.popen: Fix new `encoding` argument.

### PR DESCRIPTION
**Pull request by methane**
* Make it keyword only.
* Fix it wasn't passed to subprocess.Popen

#87901

### CODE REVIEW

**Review by hauntsaninja — APPROVED** [review]
Looks good, thanks for fixing!

File: Lib/os.py [hunk 1]
```diff
@@ -974,7 +974,7 @@ def spawnlpe(mode, file, *args):
 # command in a shell can't be supported.
 if sys.platform != 'vxworks':
     # Supply os.popen()
-    def popen(cmd, mode="r", buffering=-1, encoding=None):
+    def popen(cmd, mode="r", buffering=-1, *, encoding=None):
```
Comment (serhiy-storchaka) [inline Lib/os.py]: If add `encoding`, why not add `errors`?

**Comment by JelleZijlstra** [comment]
You should also add a unit test that uses this new parameter.

File: Lib/test/test_os.py [hunk 2]
```diff
@@ -998,7 +998,7 @@ def _empty_mapping(self):
     def test_update2(self):
         os.environ.clear()
         os.environ.update(HELLO="World")
-        with os.popen("%s -c 'echo $HELLO'" % unix_shell) as popen:
+        with os.popen("%s -c 'echo $HELLO'" % unix_shell, encoding="utf-8") as popen:
```
Comment (serhiy-storchaka) [inline Lib/test/test_os.py]: It is not a test for `os.popen`. It works without explicit `encoding`, so no need to change it.

Instead look at `test_popen`. Needed tests for `encoding` and `errors`, and they should fail if revert changes in `os.popen`, so use non-ASCII data and encodings which give different result in comparison with common locale encodings: utf-8, latin1 or cp1252.

**Review by serhiy-storchaka — COMMENTED** [review]
Is there a What's New entry?

File: Doc/library/os.rst [hunk 3]
```diff
@@ -3916,13 +3916,13 @@ written in Python, such as a mail server's external command delivery program.
    .. availability:: Unix.
 
 
-.. function:: popen(cmd, mode='r', buffering=-1, encoding=None)
+.. function:: popen(cmd, mode='r', buffering=-1, *, encoding=None, errors=None)
```
Comment (vstinner) [inline Doc/library/os.rst:3919]: Making encoding as keyword-only parameter is a backward incompatible change. Is it worth it to break old code?

File: Lib/test/test_os.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_os.py]: Would it make sense to add a new test for the private attribute to check for the encoding? To check encoding="utf-8".

File: Doc/library/os.rst (hunk 3 above)
Comment (serhiy-storchaka) [reply Doc/library/os.rst:3919]: It is a new parameter.

File: Lib/test/test_os.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_os.py]: I would prefer not to rely on private attributes if possible. Checking the result of the decoding is enough.

File: Doc/library/os.rst (hunk 3 above)
Comment (vstinner) [reply Doc/library/os.rst:3919]: My comment is about the encoding parameter which is currently a positional-or-keyword parameter. In short, I suggest:

```suggestion
.. function:: popen(cmd, mode='r', buffering=-1, encoding=None, *, errors=None)
```

Comment (JelleZijlstra) [reply Doc/library/os.rst:3919]: The `encoding` parameter was added right before 3.11b1. Making it keyword-only now seems harmless.

Comment (vstinner) [reply Doc/library/os.rst:3919]: Ah, I didn't know that it's new in Python 3.11. In that case, I'm fine with this change.

**Comment by serhiy-storchaka** [comment]
There is a dedicated test `Lib/test/test_popen.py` for this.

Test:
* Non-ASCII output with default encoding and errors.
* Non-ASCII output with different values of encoding.
* Non-ASCII output with a specified encoding and different values of errors.
* Non-ASCII command arguments and how they interact with the specified encoding.

Include tests for the locale encoding and the filesystem encoding if they are relevant.

**Comment by methane** [comment]
os.popen() is just a simple wrapper of subprocess.Popen()
And since `os.popen()` uses `shell=True` with string cmd, writing/maintain such a test in cross platform is 100x difficult than its implementation.

**Comment by methane** [comment]
Or should we revert the `encoding` argument and recommend using `subprocess` instead?

**Comment by vstinner** [comment]
> os.popen() is just a simple wrapper of subprocess.Popen()

I dislike this API since it always use a shell and so is more likely to be vulnerable to shell code injection. It may also be less efficient than avoiding a shell. I would prefer to deprecate this function, but I was never brave enough to propose deprecating it. At least, I removed platform.popen() ;-)

**Comment by methane** [comment]
I created #92836 that remove `encoding` argument and add document instead.

**Pull request closed without merging**